### DISCUSSION TITLE
pnpm monorepo: How to make fast refresh work when other sub-workspace changes？

### DISCUSSION THREAD

**Discussion by yf-yang** [vercel/next.js#57751] [Help]
### Summary

I want to know how fast refresh monitors changes of files. If someone can provide a link to the code, that would be better.

I notice https://github.com/belgattitude/nextjs-monorepo-example supports fast refresh for files other than the next app. However, I try to follow the guide but never get it worked. Then I do not know what is the key configuration that configures fast refresh's behavior.
I tried to:
   - enable `externalDir` in my next.config.js and add path aliases in tsconfig, but fast refresh does not work.
   - replace pnpm with yarn, but I cannot get yarn work for one of the packages, so I am still not sure if that is the problem.
   - set the workspaces field in `package.json` (because pnpm is not using it), but it seems it does not work.
   - change my tsconfig according to the example repo mentioned above, it does not work.

So, I am confused how fast refresh works (monitors files not in next app folder). If anyone has succeeded with pnpm or can provide some clues, I'd be grateful.

**Comment by masterbater** [comment]
If its imported it will automatically refresh, probably you need to bundle it read it from here https://turbo.build/repo/docs/handbook/publishing-packages/bundling. If you are using esm everywhere no need for bundling

**Reply by yf-yang** [reply]
Thank you for your explanation! I added more details below.
I did see the example repo uses `tsup` in its sub-workspaces, but if I got it right, when I type `pnpm dev` with the next app sub-workspace, those build scripts using `tsup` is not called, right? Specifically, do you know how to allow the next sub-workspace's bundler (webpack, or turbopack if it can do the work) to watch sub-workspace's typescript source file changes, then transpile the typescript file and trigger javascript HMR?

**Reply by yf-yang** [reply]
Sorry, I noticed that 

> If we've already set up [dev scripts](https://turbo.build/repo/docs/handbook/dev) in our turbo.json, running turbo run dev will run our packages/math dev task in parallel with our apps/web dev task.

I need to do more investigation, seems they have some relevance.

**Comment by yf-yang** [comment]
Let me add some details:
The whole repository is written with TypeScript. Now if the typescript sub-workspace get compiled, then once the javascript changes, nextjs app would reload (that's expected). However, when the typescript files are changed, nothing happens. I thought it was usual until I found in the example repo, it can capture typescript file changes. 

So actually what I am looking for is a solution that I don't need to compile all of those sub-workspaces any more. If the nextjs app dev server is up, then all the typescript files' changes in other workspaces should be watched by the bundler or somebody.
